compiler/rustc_middle/src/middle/privacy.rs

```diff
@@ -56,14 +63,47 @@ impl EffectiveVisibility {
     }
 
     pub fn is_public_at_level(&self, tag: AccessLevel) -> bool {
-        self.get(tag).map_or(false, |vis| vis.is_public())
+        self.get(tag).is_public()
+    }
+
+    pub fn map_id<OutId: Hash + Eq + Copy>(
+        &self,
+        f: impl Fn(Id) -> OutId + Copy,
+    ) -> EffectiveVisibility<OutId> {
+        EffectiveVisibility {
+            public: self.public.map_id(f),
+            exported: self.exported.map_id(f),
+            reachable: self.reachable.map_id(f),
+            reachable_from_impl_trait: self.reachable_from_impl_trait.map_id(f),
+        }
+    }
+}
+
+impl<Id: Hash + Eq + Copy + Into<DefId>> EffectiveVisibility<Id> {
+    fn update(&mut self, vis: Visibility<Id>, tag: AccessLevel, tree: impl DefIdTree) -> bool {
+        let mut changed = false;
+        for level in [
+            AccessLevel::Public,
+            AccessLevel::Exported,
+            AccessLevel::Reachable,
+            AccessLevel::ReachableFromImplTrait,
+        ] {
+            if level <= tag {
+                let current_effective_vis = self.get_mut(level);
+                if vis.is_at_least(*current_effective_vis, tree) && *current_effective_vis != vis {
```

Comment:
```suggestion
                if *current_effective_vis != vis && vis.is_at_least(*current_effective_vis, tree) {
```
The equality condition is cheaper to check.